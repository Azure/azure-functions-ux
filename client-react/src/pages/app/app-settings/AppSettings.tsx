import { Formik, FormikProps } from 'formik';
import * as React from 'react';
import { style } from 'typestyle';
import { AppSettingsFormValues } from './AppSettings.types';
import AppSettingsCommandBar from './AppSettingsCommandBar';
import AppSettingsDataLoader from './AppSettingsDataLoader';
import AppSettingsForm from './AppSettingsForm';
import { ScenarioIds } from '../../../utils/scenario-checker/scenario-ids';
import { translate, InjectedTranslateProps, TranslationFunction } from 'react-i18next';
import { useRef } from 'react';
import { ScenarioService } from '../../../utils/scenario-checker/scenario.service';

const formStyle = style({
  padding: '5px 20px',
});

const validate = (values: AppSettingsFormValues, t: TranslationFunction, scenarioChecker: ScenarioService) => {
  const duplicateDefaultDocumentsValidation = (value: string) => {
    return values.config.properties.defaultDocuments.filter(v => v === value).length > 1 ? t('fieldMustBeUnique') : null;
  };
  const hasDuplicates = array => {
    return new Set(array).size !== array.length;
  };
  const isJavaApp = values.currentlySelectedStack === 'java';
  const defaultDocumentsEnabled =
    scenarioChecker.checkScenario(ScenarioIds.defaultDocumentsSupported, { site: values.site }).status !== 'disabled';
  let hasAnyError = false;
  let errors = {
    config: {
      properties: {} as any,
    },
  };
  if (isJavaApp && !values.config.properties.javaContainer) {
    hasAnyError = true;
    errors.config.properties.javaContainer = 'required';
  }
  if (isJavaApp && !values.config.properties.javaContainerVersion) {
    hasAnyError = true;
    errors.config.properties.javaContainerVersion = 'required';
  }
  if (isJavaApp && !values.config.properties.javaVersion) {
    hasAnyError = true;
    errors.config.properties.javaVersion = 'required';
  }

  if (defaultDocumentsEnabled && hasDuplicates(values.config.properties.defaultDocuments)) {
    hasAnyError = true;
    errors.config.properties.defaultDocuments = values.config.properties.defaultDocuments.map(value => {
      return duplicateDefaultDocumentsValidation(value);
    });
  }

  return hasAnyError ? errors : {};
};

const AppSettings: React.SFC<InjectedTranslateProps> = props => {
  const { t } = props;
  const scenarioCheckerRef = useRef(new ScenarioService(t));
  const scenarioChecker = scenarioCheckerRef.current!;
  return (
    <AppSettingsDataLoader>
      {({ initialFormValues, saving, loading, onSubmit }) =>
        (console.log(initialFormValues) as any) || (
          <Formik
            initialValues={initialFormValues}
            onSubmit={onSubmit}
            enableReinitialize={true}
            validate={values => validate(values, t, scenarioChecker)}
            validateOnBlur={false}
            validateOnChange={false}>
            {(formProps: FormikProps<AppSettingsFormValues>) => (
              <form>
                <AppSettingsCommandBar
                  submitForm={formProps.submitForm}
                  resetForm={formProps.resetForm}
                  disabled={!formProps.values.siteWritePermission || saving || loading}
                  dirty={formProps.dirty}
                />
                <div className={formStyle}>
                  <AppSettingsForm {...formProps} />
                </div>
              </form>
            )}
          </Formik>
        )
      }
    </AppSettingsDataLoader>
  );
};

export default translate('translation')(AppSettings);
