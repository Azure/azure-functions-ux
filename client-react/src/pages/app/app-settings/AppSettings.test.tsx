import { shallow, ShallowWrapper } from 'enzyme';
import * as React from 'react';
import { AppSettings } from './AppSettings';
import mockState from '../../../mocks/mockState';
describe('App Settings', () => {
  describe('Shallow Testing', () => {
    let container: ShallowWrapper<any, any, AppSettings>;
    const fetchSite = jest.fn();
    const fetchSettings = jest.fn();
    const fetchConfig = jest.fn();
    const fetchCNStrings = jest.fn();
    const fetchStacks = jest.fn();
    const updateConfig = jest.fn();
    const updateSite = jest.fn();

    beforeEach(() => {
      container = shallow(
        <AppSettings
          fetchSite={fetchSite}
          fetchSettings={fetchSettings}
          fetchConfig={fetchConfig}
          fetchConnStrings={fetchCNStrings}
          fetchStacks={fetchStacks}
          site={mockState.site.site}
          appSettings={mockState.appSettings.settings}
          virtualApplications={mockState.webConfig.virtualApplications}
          config={mockState.webConfig.config}
          connectionStrings={mockState.connectionStrings.connectionStrings}
          currentlySelectedStack={mockState.webConfig.currentlySelectedStack}
          updateConfig={updateConfig}
          updateSite={updateSite}
        />
      );
    });

    afterEach(() => {
      updateSite.mockReset();
      updateConfig.mockReset();
    });
    it('renders without creashing', () => {
      expect(container.length).toEqual(1);
    });

    it('renders correctly', () => {
      expect(container).toMatchSnapshot('app-settings-form');
    });
    it('dispatches fetch calls on load', () => {
      expect(fetchSite).toBeCalled();
      expect(fetchSettings).toBeCalled();
      expect(fetchConfig).toBeCalled();
      expect(fetchCNStrings).toBeCalled();
      expect(fetchStacks).toBeCalled();
    });

    it('submit calls update site and update config', () => {
      expect(updateSite).not.toHaveBeenCalled();
      expect(updateConfig).not.toHaveBeenCalled();
      const values = container.instance().initialValues(container.props());
      container.instance().onSubmit(values, { setSubmitting: jest.fn() } as any);
      expect(updateSite).toHaveBeenCalled();
      expect(updateConfig).toHaveBeenCalled();
    });

    it('submit calls set submitting to true while submitting', () => {
      const values = container.instance().initialValues(container.props());
      container.instance().onSubmit(values, { setSubmitting: jest.fn() } as any);
      expect(container.state('isSubmitting')).toBeTruthy();
    });

    it('isSubmitting is set to false when completed', async () => {
      const values = container.instance().initialValues(container.props());
      updateSite.mockResolvedValue(null);
      updateConfig.mockResolvedValue(null);
      await container.instance().onSubmit(values, { setSubmitting: jest.fn() } as any);
      expect(container.state('isSubmitting')).toBeFalsy();
    });
  });
});
