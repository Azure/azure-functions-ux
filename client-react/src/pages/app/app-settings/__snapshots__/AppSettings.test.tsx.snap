// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Settings Shallow Testing renders correctly: app-settings-form 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AppSettings
    appSettings={Array []}
    config={
      Object {
        "id": "/subscriptions/subid/resourceGroups/rg/providers/Microsoft.Web/sites/siteName/config/web",
        "location": "West US",
        "name": "siteName",
        "properties": Object {
          "alwaysOn": false,
          "appCommandLine": "",
          "autoHealEnabled": false,
          "azureStorageAccounts": Object {},
          "customAppPoolIdentityAdminState": false,
          "customAppPoolIdentityTenantState": false,
          "defaultDocuments": Array [
            "Default.htm",
            "Default.html",
            "Default.asp",
            "index.htm",
            "index.html",
            "iisstart.htm",
            "default.aspx",
            "index.php",
            "hostingstart.html",
          ],
          "detailedErrorLoggingEnabled": false,
          "ftpsState": "AllAllowed",
          "http20Enabled": false,
          "httpLoggingEnabled": false,
          "linuxFxVersion": "",
          "loadBalancing": 1,
          "localMySqlEnabled": false,
          "logsDirectorySizeLimit": 35,
          "managedPipelineMode": 0,
          "minTlsVersion": "1.2",
          "netFrameworkVersion": "v4.0",
          "nodeVersion": "",
          "numberOfWorkers": 1,
          "phpVersion": "5.6",
          "publishingUsername": "$siteName",
          "pythonVersion": "",
          "remoteDebuggingEnabled": false,
          "requestTracingEnabled": false,
          "reservedInstanceCount": 0,
          "scmType": "None",
          "siteAuthEnabled": false,
          "use32BitWorkerProcess": true,
          "vnetName": "",
          "webSocketsEnabled": false,
          "winAuthAdminState": 0,
          "winAuthTenantState": 0,
          "windowsFxVersion": "",
        },
        "type": "Microsoft.Web/sites/config",
      }
    }
    connectionStrings={Array []}
    currentlySelectedStack="dotnet"
    fetchConfig={
      [MockFunction] {
        "calls": Array [
          Array [],
          Array [],
        ],
      }
    }
    fetchConnStrings={
      [MockFunction] {
        "calls": Array [
          Array [],
          Array [],
        ],
      }
    }
    fetchSettings={
      [MockFunction] {
        "calls": Array [
          Array [],
          Array [],
        ],
      }
    }
    fetchSite={
      [MockFunction] {
        "calls": Array [
          Array [],
          Array [],
        ],
      }
    }
    fetchStacks={
      [MockFunction] {
        "calls": Array [
          Array [],
          Array [],
        ],
      }
    }
    site={
      Object {
        "id": "/subscriptions/subid/resourceGroups/rg/providers/Microsoft.Web/sites/siteName",
        "kind": "app",
        "location": "West US",
        "name": "siteName",
        "properties": Object {
          "adminEnabled": true,
          "availabilityState": 0,
          "clientAffinityEnabled": true,
          "clientCertEnabled": false,
          "containerSize": 0,
          "contentAvailabilityState": 0,
          "dailyMemoryTimeQuota": 0,
          "defaultHostName": "siteName.azurewebsites.net",
          "deploymentId": "siteName",
          "enabled": true,
          "enabledHostNames": Array [
            "siteName.azurewebsites.net",
            "siteName.scm.azurewebsites.net",
          ],
          "homeStamp": "waws-prod-bay-061",
          "hostNameSslStates": Array [
            Object {
              "hostType": 0,
              "ipBasedSslState": 0,
              "name": "siteName.azurewebsites.net",
              "sslState": 0,
            },
            Object {
              "hostType": 1,
              "ipBasedSslState": 0,
              "name": "siteName.scm.azurewebsites.net",
              "sslState": 0,
            },
          ],
          "hostNames": Array [
            "siteName.azurewebsites.net",
          ],
          "hostNamesDisabled": false,
          "httpsOnly": false,
          "isXenon": false,
          "kind": "app",
          "lastModifiedTimeUtc": "2018-10-14T06:37:18.417",
          "name": "siteName",
          "outboundIpAddresses": "40.118.255.137,40.118.255.231,40.118.250.218,40.118.252.35",
          "possibleOutboundIpAddresses": "40.118.255.137,40.118.255.231,40.118.250.218,40.118.252.35,40.118.255.137,40.118.255.231,40.118.250.218,40.118.252.35",
          "repositorySiteName": "siteName",
          "reserved": false,
          "resourceGroup": "rg",
          "runtimeAvailabilityState": 0,
          "scmSiteAlsoStopped": false,
          "selfLink": "https://waws-prod-bay-061.api.azurewebsites.windows.net:454/subscriptions/subid/webspaces/ZUMO-Not-Delete-WestUSwebspace/sites/siteName",
          "serverFarmId": "/subscriptions/subid/resourceGroups/ZUMO-Not-Delete/providers/Microsoft.Web/serverfarms/DefaultTier1ServerFarm",
          "siteDisabledReason": 0,
          "siteProperties": Object {
            "properties": Array [
              Object {
                "name": "LinuxFxVersion",
                "value": "",
              },
              Object {
                "name": "WindowsFxVersion",
                "value": "",
              },
            ],
          },
          "sku": "Standard",
          "state": "Running",
          "storageRecoveryDefaultState": "Running",
          "usageState": 0,
          "webSpace": "",
        },
        "type": "Microsoft.Web/sites",
      }
    }
    t={[Function]}
    updateConfig={[MockFunction]}
    updateSite={[MockFunction]}
    virtualApplications={
      Array [
        Object {
          "physicalPath": "site\\\\wwwroot",
          "preloadEnabled": false,
          "virtualDirectories": Array [],
          "virtualDirectory": false,
          "virtualPath": "/",
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": [Function],
      "enableReinitialize": true,
      "initialValues": Object {
        "appSettings": Array [],
        "config": Object {
          "id": "/subscriptions/subid/resourceGroups/rg/providers/Microsoft.Web/sites/siteName/config/web",
          "location": "West US",
          "name": "siteName",
          "properties": Object {
            "alwaysOn": false,
            "appCommandLine": "",
            "autoHealEnabled": false,
            "azureStorageAccounts": Object {},
            "customAppPoolIdentityAdminState": false,
            "customAppPoolIdentityTenantState": false,
            "defaultDocuments": Array [
              "Default.htm",
              "Default.html",
              "Default.asp",
              "index.htm",
              "index.html",
              "iisstart.htm",
              "default.aspx",
              "index.php",
              "hostingstart.html",
            ],
            "detailedErrorLoggingEnabled": false,
            "ftpsState": "AllAllowed",
            "http20Enabled": false,
            "httpLoggingEnabled": false,
            "linuxFxVersion": "",
            "loadBalancing": 1,
            "localMySqlEnabled": false,
            "logsDirectorySizeLimit": 35,
            "managedPipelineMode": 0,
            "minTlsVersion": "1.2",
            "netFrameworkVersion": "v4.0",
            "nodeVersion": "",
            "numberOfWorkers": 1,
            "phpVersion": "5.6",
            "publishingUsername": "$siteName",
            "pythonVersion": "",
            "remoteDebuggingEnabled": false,
            "requestTracingEnabled": false,
            "reservedInstanceCount": 0,
            "scmType": "None",
            "siteAuthEnabled": false,
            "use32BitWorkerProcess": true,
            "vnetName": "",
            "webSocketsEnabled": false,
            "winAuthAdminState": 0,
            "winAuthTenantState": 0,
            "windowsFxVersion": "",
          },
          "type": "Microsoft.Web/sites/config",
        },
        "connectionStrings": Array [],
        "currentlySelectedStack": "dotnet",
        "site": Object {
          "id": "/subscriptions/subid/resourceGroups/rg/providers/Microsoft.Web/sites/siteName",
          "kind": "app",
          "location": "West US",
          "name": "siteName",
          "properties": Object {
            "adminEnabled": true,
            "availabilityState": 0,
            "clientAffinityEnabled": true,
            "clientCertEnabled": false,
            "containerSize": 0,
            "contentAvailabilityState": 0,
            "dailyMemoryTimeQuota": 0,
            "defaultHostName": "siteName.azurewebsites.net",
            "deploymentId": "siteName",
            "enabled": true,
            "enabledHostNames": Array [
              "siteName.azurewebsites.net",
              "siteName.scm.azurewebsites.net",
            ],
            "homeStamp": "waws-prod-bay-061",
            "hostNameSslStates": Array [
              Object {
                "hostType": 0,
                "ipBasedSslState": 0,
                "name": "siteName.azurewebsites.net",
                "sslState": 0,
              },
              Object {
                "hostType": 1,
                "ipBasedSslState": 0,
                "name": "siteName.scm.azurewebsites.net",
                "sslState": 0,
              },
            ],
            "hostNames": Array [
              "siteName.azurewebsites.net",
            ],
            "hostNamesDisabled": false,
            "httpsOnly": false,
            "isXenon": false,
            "kind": "app",
            "lastModifiedTimeUtc": "2018-10-14T06:37:18.417",
            "name": "siteName",
            "outboundIpAddresses": "40.118.255.137,40.118.255.231,40.118.250.218,40.118.252.35",
            "possibleOutboundIpAddresses": "40.118.255.137,40.118.255.231,40.118.250.218,40.118.252.35,40.118.255.137,40.118.255.231,40.118.250.218,40.118.252.35",
            "repositorySiteName": "siteName",
            "reserved": false,
            "resourceGroup": "rg",
            "runtimeAvailabilityState": 0,
            "scmSiteAlsoStopped": false,
            "selfLink": "https://waws-prod-bay-061.api.azurewebsites.windows.net:454/subscriptions/subid/webspaces/ZUMO-Not-Delete-WestUSwebspace/sites/siteName",
            "serverFarmId": "/subscriptions/subid/resourceGroups/ZUMO-Not-Delete/providers/Microsoft.Web/serverfarms/DefaultTier1ServerFarm",
            "siteDisabledReason": 0,
            "siteProperties": Object {
              "properties": Array [
                Object {
                  "name": "LinuxFxVersion",
                  "value": "",
                },
                Object {
                  "name": "WindowsFxVersion",
                  "value": "",
                },
              ],
            },
            "sku": "Standard",
            "state": "Running",
            "storageRecoveryDefaultState": "Running",
            "usageState": 0,
            "webSpace": "",
          },
          "type": "Microsoft.Web/sites",
        },
        "virtualApplications": Array [
          Object {
            "physicalPath": "site\\\\wwwroot",
            "preloadEnabled": false,
            "virtualDirectories": Array [],
            "virtualDirectory": false,
            "virtualPath": "/",
          },
        ],
      },
      "isInitialValid": false,
      "onSubmit": [Function],
      "validateOnBlur": true,
      "validateOnChange": true,
    },
    "ref": null,
    "rendered": [Function],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": [Function],
        "enableReinitialize": true,
        "initialValues": Object {
          "appSettings": Array [],
          "config": Object {
            "id": "/subscriptions/subid/resourceGroups/rg/providers/Microsoft.Web/sites/siteName/config/web",
            "location": "West US",
            "name": "siteName",
            "properties": Object {
              "alwaysOn": false,
              "appCommandLine": "",
              "autoHealEnabled": false,
              "azureStorageAccounts": Object {},
              "customAppPoolIdentityAdminState": false,
              "customAppPoolIdentityTenantState": false,
              "defaultDocuments": Array [
                "Default.htm",
                "Default.html",
                "Default.asp",
                "index.htm",
                "index.html",
                "iisstart.htm",
                "default.aspx",
                "index.php",
                "hostingstart.html",
              ],
              "detailedErrorLoggingEnabled": false,
              "ftpsState": "AllAllowed",
              "http20Enabled": false,
              "httpLoggingEnabled": false,
              "linuxFxVersion": "",
              "loadBalancing": 1,
              "localMySqlEnabled": false,
              "logsDirectorySizeLimit": 35,
              "managedPipelineMode": 0,
              "minTlsVersion": "1.2",
              "netFrameworkVersion": "v4.0",
              "nodeVersion": "",
              "numberOfWorkers": 1,
              "phpVersion": "5.6",
              "publishingUsername": "$siteName",
              "pythonVersion": "",
              "remoteDebuggingEnabled": false,
              "requestTracingEnabled": false,
              "reservedInstanceCount": 0,
              "scmType": "None",
              "siteAuthEnabled": false,
              "use32BitWorkerProcess": true,
              "vnetName": "",
              "webSocketsEnabled": false,
              "winAuthAdminState": 0,
              "winAuthTenantState": 0,
              "windowsFxVersion": "",
            },
            "type": "Microsoft.Web/sites/config",
          },
          "connectionStrings": Array [],
          "currentlySelectedStack": "dotnet",
          "site": Object {
            "id": "/subscriptions/subid/resourceGroups/rg/providers/Microsoft.Web/sites/siteName",
            "kind": "app",
            "location": "West US",
            "name": "siteName",
            "properties": Object {
              "adminEnabled": true,
              "availabilityState": 0,
              "clientAffinityEnabled": true,
              "clientCertEnabled": false,
              "containerSize": 0,
              "contentAvailabilityState": 0,
              "dailyMemoryTimeQuota": 0,
              "defaultHostName": "siteName.azurewebsites.net",
              "deploymentId": "siteName",
              "enabled": true,
              "enabledHostNames": Array [
                "siteName.azurewebsites.net",
                "siteName.scm.azurewebsites.net",
              ],
              "homeStamp": "waws-prod-bay-061",
              "hostNameSslStates": Array [
                Object {
                  "hostType": 0,
                  "ipBasedSslState": 0,
                  "name": "siteName.azurewebsites.net",
                  "sslState": 0,
                },
                Object {
                  "hostType": 1,
                  "ipBasedSslState": 0,
                  "name": "siteName.scm.azurewebsites.net",
                  "sslState": 0,
                },
              ],
              "hostNames": Array [
                "siteName.azurewebsites.net",
              ],
              "hostNamesDisabled": false,
              "httpsOnly": false,
              "isXenon": false,
              "kind": "app",
              "lastModifiedTimeUtc": "2018-10-14T06:37:18.417",
              "name": "siteName",
              "outboundIpAddresses": "40.118.255.137,40.118.255.231,40.118.250.218,40.118.252.35",
              "possibleOutboundIpAddresses": "40.118.255.137,40.118.255.231,40.118.250.218,40.118.252.35,40.118.255.137,40.118.255.231,40.118.250.218,40.118.252.35",
              "repositorySiteName": "siteName",
              "reserved": false,
              "resourceGroup": "rg",
              "runtimeAvailabilityState": 0,
              "scmSiteAlsoStopped": false,
              "selfLink": "https://waws-prod-bay-061.api.azurewebsites.windows.net:454/subscriptions/subid/webspaces/ZUMO-Not-Delete-WestUSwebspace/sites/siteName",
              "serverFarmId": "/subscriptions/subid/resourceGroups/ZUMO-Not-Delete/providers/Microsoft.Web/serverfarms/DefaultTier1ServerFarm",
              "siteDisabledReason": 0,
              "siteProperties": Object {
                "properties": Array [
                  Object {
                    "name": "LinuxFxVersion",
                    "value": "",
                  },
                  Object {
                    "name": "WindowsFxVersion",
                    "value": "",
                  },
                ],
              },
              "sku": "Standard",
              "state": "Running",
              "storageRecoveryDefaultState": "Running",
              "usageState": 0,
              "webSpace": "",
            },
            "type": "Microsoft.Web/sites",
          },
          "virtualApplications": Array [
            Object {
              "physicalPath": "site\\\\wwwroot",
              "preloadEnabled": false,
              "virtualDirectories": Array [],
              "virtualDirectory": false,
              "virtualPath": "/",
            },
          ],
        },
        "isInitialValid": false,
        "onSubmit": [Function],
        "validateOnBlur": true,
        "validateOnChange": true,
      },
      "ref": null,
      "rendered": [Function],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
